### ISSUE TITLE
Phyloseq CCA: 'x' and 'w' must have the same length

### ISSUE THREAD

**Issue by pauGuas** [vegandevs/vegan#459]
Hi, I am trying to run a CCA on the numerical environmental factors that influence microbiome composition (phyloseq object).  The environmental factors are summarized in "var". When I run the code below, I get "Error in weighted.mean.default(newX[, i], ...) : 'x' and 'w' must have the same length".  I don't know what 'w' is in this.  Does anyone know what to do with this? Thank you!

`ps <- phyloseq_object
prop= transform_sample_counts(ps, function(OTU) OTU/sum(OTU))
prop_df= as.data.frame(otu_table(prop))
prop_dft= as.data.frame(t(prop_df))

prop_dft <- ps_prop_dft[rowSums(prop_dft[, -1])>0, ]

var <- c("Chloride", "Inorganic.N", "Kjeldahl.N",  "Dissolved.O", "Temp",  "TSS", "Velocity.of.Discharge", "pH")

ccamodel= cca(prop_dft ~ var)`

**Comment by MarkReader** [comment]
The "w" field will be, I guess, the Population Weights. One value (weight) for each n (in sample), hence w has to be same length as your obs. They will be different if they are population weights, such as (often) much higher weights for known demographics that are undersampled (higher n/r rates, etc). If you just want the arithmetic (sample, not population) mean, just set them all to 1.000. :)

**Comment by gavinsimpson** [comment]
@MarkReader CCA involves abundance-weighting the data so the error could just as easily be being raised from functions internal to `cca`.

@pauGuas Solve the obvious problem first; that's not how you fit a CCA model. The RHS of the formula should be a standard R formula: ` ~ x1 + x2 + x3` etc, not a character vector.

If you want to proceed using the defined `var`, use:

```r
ord <- cca(prop_dft ~ ., data = ENV[, var])
```

replacing `ENV` with your data frame containing the environmental variables that you want to use in the model.

**Comment by MarkReader** [comment]
Very likely so. (But, out of curiosity @gavinsimpson what are the "w"? Will be really grateful if you can DM me, if you like.).

**Comment by jarioksa** [comment]
Yes, the error is triggered in an internal function: user should not and cannot supply weights to `cca`. If you are interested in details, look at the function `vegan:::ordConstrain` which has the following code block:
```r
    if (!is.null(RW)) {
        envcentre <- apply(X, 2, weighted.mean, w = RW)
        X <- scale(X, center = envcentre, scale = FALSE)
        X <- sweep(X, 1, sqrt(RW), "*")
    }
```
The weights `RW` are the row weights of the dependent data used as internal weights in Correspondence Analysis (the method, irrespective of its software implementation). These weights are calculated in `vegan:::initCA` function.

However, I cannot understand how you reach this error. Like @gavinsimpson , the model definition is wrong. Moreover, we do check the dimensions before going to `vegan:::ordConstrained`, and the commands in the original message should give this error message:
```r
Error in ordConstrained(X, Y, Z, method = "cca") : 
  dependent data and constraints must have the same number of rows
```
(and I got this when I tried to mimick the  original commands – I could not reproduce them as there was nothing reproducible.) The command should only work if `length(var) == NROW(prop_dft)`, and then you would get a useless result identical to unconstrained CA (as the number of levels of `var` factor equals the number of observations).

The way to proceed is to follow Gav's instruction, and if it still fails, provide a reproducible example (where "reproducible" means something we can run and get the same error).
